### PR TITLE
Add library for managing information on ManagedPool tokens

### PR DESCRIPTION
**Pull request by TomAFrench**
This PR adds `ManagedPoolTokenLib` which handles all data operations to do with the `_tokenState` mapping, i.e. it encapsulates token scaling factors and weights. As the data layout is the same as current, I've modified ManagedPool to make use of this new library.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Looking into test case failure.

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol [hunk 1]
```diff
@@ -573,10 +564,13 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard, ICo
         // is achieved efficiently by simply updating the sum of the denormalized weights.
         _denormWeightSum = weightSumAfterAdd;
 
-        // Finally, we store the new token's weight.
-        // `_encodeTokenState` performs an external call to `token` (to get its decimals), but this is
-        // reentrancy safe since that is a static call.
-        _tokenState[token] = _encodeTokenState(token, normalizedWeight, normalizedWeight, weightSumAfterAdd);
+        // Finally, we store the new token's weight and scaling factor.
+        _tokenState[token] = ManagedPoolTokenLib.setTokenWeight(
+            ManagedPoolTokenLib.setTokenScalingFactor(bytes32(0), token),
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/ManagedPool.sol:569]: Sneakily initializing the state to include the scaling factor is rather odd - I had to spend about 3 minutes going back and forth trying to understand what this line was doing.

An issue here is that there's no guarantee the full state is scaling factor + weight. We should perhaps have a function that returns a fully initialied state, so we can be sure we're not forgetting anything.

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol [hunk 2]
```diff
@@ -708,7 +695,7 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard, ICo
         // all other token weights accordingly.
         // Clean up data structures and update the token count
         delete _tokenState[token];
-        _denormWeightSum -= _denormalizeWeight(tokenNormalizedWeight, _denormWeightSum);
+        _denormWeightSum -= tokenNormalizedWeight.mulUp(_denormWeightSum);
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/ManagedPool.sol:698]: DId you feel these didn't add much?

File: pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/openzeppelin/IERC20.sol";
+
+import "../managed/ManagedPoolTokenLib.sol";
+
+contract MockManagedPoolTokenLib {
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol:22]: If only there was a way to test libraries directly from Solidity

File: pkg/pool-weighted/test/WeightedPool.test.ts [hunk 4]
```diff
@@ -62,7 +62,7 @@ describe('WeightedPool', function () {
 
         // Perform a recovery mode exit. This will reduce the invariant but this isn't tracked due to recovery mode.
         const preExitInvariant = await pool.getLastPostJoinExitInvariant();
-        await pool.recoveryModeExit({ from: lp, bptIn: fp(100) });
+        await pool.recoveryModeExit({ from: lp, bptIn: fp(10) });
```
Comment (nventuro) [inline pkg/pool-weighted/test/WeightedPool.test.ts:65]: Why this change?

File: pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    const clearedResult = mask.and(result);
+    const clearedWord = mask.and(word);
+    expect(clearedResult).to.equal(clearedWord);
+  }
+
+  describe('token scaling factor', () => {
+    const DECIMAL_DIFF_OFFSET = 128;
+    const DECIMAL_DIFF_WIDTH = 5;
+
+    async function assertTokenScalingFactor(
```
Comment (nventuro) [inline pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts:34]: I'd have this function also receive an `expected` argument, and also check against that.

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        returns (uint256, uint256)
+    {
+        return (
+            _decodeWeight(tokenState.decodeUint(_START_DENORM_WEIGHT_OFFSET, _DENORM_WEIGHT_WIDTH), denormWeightSum),
+            _decodeWeight(tokenState.decodeUint(_END_DENORM_WEIGHT_OFFSET, _DENORM_WEIGHT_WIDTH), denormWeightSum)
+        );
+    }
+
+    function getMinimumTokenEndWeight(
+        mapping(IERC20 => bytes32) storage tokenStates,
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:93]: Yeah, I toyed with the idea of building the array of `tokenStates` inside `ManagedPool` but extra memory cost + and avoids `ManagedPool` from having to think as much.

File: pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol (hunk 3 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol:22]: I haven't found rly anything that would allow that.

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    // Getters
+
+    /**
+     * @notice Returns the token's scaling factor.
+     */
+    function getTokenScalingFactor(bytes32 tokenState) internal pure returns (uint256) {
+        uint256 decimalsDifference = tokenState.decodeUint(_DECIMAL_DIFF_OFFSET, _DECIMAL_DIFF_WIDTH);
+
+        // This is equivalent to `10**(18+decimalsDifference)` but this form optimizes for 18 decimal tokens.
+        return FixedPoint.ONE * 10**decimalsDifference;
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:56]: It's an EXP opcode which has a cost of `10 + 50 * exponent_byte_size`. I was thinking that we could do the multiplication for free if we shove `FixedPoint.ONE` up in the exponent but then we'd always have a non-zero exponent and pay 50 rather than 10 gas for most tokens.

File: pkg/pool-weighted/test/WeightedPool.test.ts (hunk 4 above)
Comment (TomAFrench) [reply pkg/pool-weighted/test/WeightedPool.test.ts:65]: The tests were failing due to `lp` not having enough BPT

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol (hunk 2 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPool.sol:698]: Through offloading most of the usages into `ManagedPoolTokenLib` this ended up being the only place where `_denormalizeWeight` was used. I think at this point inlining it makes sense.

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+     * @notice Updates a token's start and end weights.
+     * @dev To be called when initiating a gradual weight change to update the two values to interpolate between.
+     */
+    function setTokenWeight(
+        bytes32 tokenState,
+        uint256 normalizedStartWeight,
+        uint256 normalizedEndWeight,
+        uint256 denormWeightSum
+    ) internal pure returns (bytes32) {
+        return
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:128]: I think that times in which we write new state are going to be fairly localised and only change a single field. This means that we can merge the modification and setting of state.

```
bytes32 state = _tokenState[token];
/* calculations of start and end weights */
_tokenState[token] =  ManagedPoolTokenLib.setTokenWeight(state, newStartWeight, newEndWeight, denormWeightSum); 
```

The chance of missing a write only occurs if we need to perform multiple updates to the token state while reading values from a half-updated state which I don't think would ever happen.

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    // Denormalized weights are stored using the WeightCompression library as a percentage of the maximum absolute
+    // denormalized weight: independent of the current _denormWeightSum, which avoids having to recompute the denorm
+    // weights as the sum changes.
+    uint256 private constant _MAX_DENORM_WEIGHT = 1e22; // FP 10,000
+
+    // Store scaling factor and start/end denormalized weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // [ 123 bits |  5 bits  |  64 bits   |   64 bits    |
+    // [ unused   | decimals | end denorm | start denorm |
+    // |MSB                                           LSB|
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:39]: This library does require some knowledge of the `denormWeightSum` as it takes it as an argument to normalize weights. Agreed on needing to update comments.

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:128]: If I had to chose between the two approaches though I would definitely go b).

File: pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts (hunk 5 above)
Comment (TomAFrench) [reply pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts:34]: As in an expected value for `result`? I'm inclined to work just off of observable values instead as this is all the Pool cares about (the pool should only use `_tokenState` to feed it into this lib).

Comment (TomAFrench) [reply pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts:34]: This would likely get pretty yucky as we'd have to be able to predict the rounding errors involved in compressing the weights as we can't do fuzzy matching on the entire state without decoding it first (at which point we'd just be back to the current testing regime but we're testing everything in every test.)

**Comment by TomAFrench** [comment]
Note that I've expanded `_MAX_DENORM_WEIGHT` from `1e22` to `2**64`. I'm not sure why we were not using the full range beforehand.

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol (hunk 1 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPool.sol]: [95bcb3b](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1730/commits/95bcb3bac511e86250b6a9862572c5ce0b6661ae)

File: pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:162]: Shall we wait until we have `ManagedBasePool` and start flattening this to decide how to move down this path?

[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol:162]: We decided to go with alternative c) leave everything as is, since the arguments presented above are good.

File: pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol (hunk 3 above)
Comment (nventuro) [reply pkg/pool-weighted/contracts/test/MockManagedPoolTokenLib.sol:22]: *standing ovation*

File: pkg/pool-weighted/test/WeightedPool.test.ts (hunk 4 above)
Comment (nventuro) [reply pkg/pool-weighted/test/WeightedPool.test.ts:65]: But why did the tests start failing? Did we change anything else?

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol [hunk 10]
```diff
@@ -210,6 +200,13 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard, ICo
 
         _setManagementAumFeePercentage(params.managementAumFeePercentage);
 
+        // Write the scaling factors for each token into their token state.
+        // We do this before setting the weights in `_startGradualWeightChange` so we start from a empty token state.
+        for (uint256 i = 0; i < totalTokens; i++) {
+            IERC20 token = params.tokens[i];
+            _tokenState[token] = ManagedPoolTokenLib.setTokenScalingFactor(bytes32(0), token);
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/ManagedPool.sol:207]: Bit odd that we don't use the initialization function here

**Review by nventuro — APPROVED** [review]

File: pkg/pool-weighted/test/WeightedPool.test.ts (hunk 4 above)
Comment (TomAFrench) [reply pkg/pool-weighted/test/WeightedPool.test.ts]: I've reverted the change and it seems to work again. I can't reproduce locally.

File: pkg/pool-weighted/contracts/managed/ManagedPool.sol (hunk 10 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPool.sol:207]: Will address separately.
